blog/_posts/2017-09-06-gsoc-derivative_operators.md

```diff
@@ -0,0 +1,156 @@
+
+## MOTIVATION
+The general idea of finite difference methods is to generate finite difference weights corresponding to a differential operator allowing a certain level of approximation. The time and space variable are divided to form a grid where 
+$h = \Delta x = \frac{1}{N+1}$ and $x_i = ih$ for $i = 0, 1,...,N+1$ and 
+$k = \Delta t = \frac{T}{M+1}$ and $t_j = jk$ for $j = 0, 1,...,M+1$. \\ 
+The discrete unknowns are scalars $u_i^j$ for the above values of i and j, and it is hoped that $u_i^j$ will be an approximation of $u(x_i, t_j)$. The right-hand side of the equation is discretized by setting $f_i^j = f(x_i, t_j)$. We also use the notation
+
+$$ U^{j} = \begin{pmatrix}
+u_1^j\\ 
+u_2^j \\
+\vdots \\
+u_N^j
+\end{pmatrix} $$
+to denote the vector of approximate values on the space grid at time $t_j$.
+
+$$\frac{\partial u}{\partial t}(x_i,t_j) \approx \frac{u(x_i,t_{j+1}) - u(x_i,t_j)}{k}$$
+
+and for the second space derivative, by combining a forward and a backward differential quotient, we obtain the central approximation
+
+$$ \frac{\partial^2u   }{\partial x^2}(x_i,t_j) \approx \frac{\frac{u(x_{i+1},t_j) - u(x_i,t_j)}{h} - \frac{u(x_i,t_j) - u(x_{i-1},t_j)}{h}}{h} = \frac{u(x_{i-1},t_j) - 2u(x_i,t_j) + u(x_{i+1}, t_j)}
+{h^2} $$
+
+So the weights corresponding to the second order partial derivative are $[1, -2, 1]$. The finite difference method mimics these approximations by replacing the exact values of the solution at the grid points by the discrete unknowns.
+
+In this particular case, we end up with the following scheme:
+
+$$
+\left\{\begin{matrix}
+\frac{u_j^{j+1} - u_i^j}{k} - \frac{u_{i-1}^j - 2u_i^j + u_{i+1}^j}{h^2} = f_i^j \textit{ for } i = 1,...,N,j = 1,...,M \\ 
+u_i^0 = u_0(x_i) \textit{ for } i = 1,...,N\\ 
+u_0^j = u_{N+1}^j = 0 \textit{ for } j = 1,...,M+1\\
+\end{matrix}\right.
+$$
+
+This is the stencil rewritten as a recurrence. Writing it out in vector form, we get:-
+$\frac{U^{j+1} - U^j}{k} + A_h U^j = F^j \textit{ for } j = 1,...,M$
+
+When we want to apply this operator on the vector $U$, the weight vector turns into a matrix called the transformation matrix $A_h$
+
+$$ A_h = \begin{pmatrix}
+2 & -1 & 0 &  \cdots  & 0\\ 
+-1 & 2 & -1 &  \cdots  & 0\\ 
+\vdots & \ddots & \ddots & \ddots & \vdots \\ 
+0 & \cdots & -1 & 2  & -1 \\ 
+0 & \cdots & 0 & -1 & 2 \\
+\end{pmatrix}
+\textit{such that }
+\frac{\partial^2U^j}{\partial x^2} \approx A_h*U^j $$
+
+But matrix multiplication is costly, therefore it would be preferable to have the linear operator of the double partial differential instead of the transformation matrix.
+It would look something like:-
+
+    function double_partial(x,dx)
+      for i in 2:length(dx)-1
+        dx[i] = -1*x[i-1] + 2*x[i] + -1*x[i+1]
+      end
+      dx[1] = 2*x[1] + -1*x[2]
+      dx[end] = -1*x[end-1] + 2*x[end]
+    end
```

Comment:
Might want to mention that this is O(n) operations as opposed to O(n^2), while still avoiding the overheads of sparse matrices.